### PR TITLE
feat(openclaw): plugin-first memory parity with shared policy

### LINKED ISSUE #68: feat: Overhaul OpenClaw integration — MCP-first approach, plugin option, natural language commands
## Context

Discord feedback from BasuruK and community members surfaced several friction points with the current OpenClaw integration. The current setup is **curl-only via a skill file** — most users find this clunky and report the agent inconsistently uses curl commands, preferring MCP tools when available.

Jack's take: rewrite the skill to work *with* MCP instead of fighting it, and explore an OpenClaw plugin as a bundled solution.

## Current State

- **Skill** (`templates/openclaw/skill/SKILL.md`): curl-based API reference, behavioral rules, importance levels
- **CLI** (`src/cli/openclaw.ts`): installs skill to `~/.openclaw/skills/automem/`, configures env vars in `openclaw.json`
- **Setup docs** (`templates/openclaw/OPENCLAW_SETUP.md`): points to curl architecture exclusively
- **No MCP option** for OpenClaw at all
- **No OpenClaw plugin** (https://docs.openclaw.ai/tools/plugin)
- **No natural language commands** (`remember X`, `what do you know about X`)

### From BasuruK (power user running AutoMem + memory-core)

1. **Agent doesn't consistently use curl from skill** — prefers MCP tools when available. Explicit instructions in skill help but aren't reliable. The agent itself explained why: no native tool integration, manual JSON payload construction is clunky vs typed MCP tools.
2. **No auto-detection** — importance levels are defined but there's no trigger mechanism. Agent has to "remember" to store, so important context gets lost.
3. **No natural user interaction** — can't say "remember that I prefer terse responses" and have it persist. Need to explicitly ask agent to run curl.
4. **Memory suggestion system** — would be cool if agent got nudged: "Hey, want to store this?" when something significant happens.

### From George-RD

- Prefers **skill over MCP** — finds it more future-proof and easier to tweak/adjust.
- Important counterpoint: we should still support skill-based usage, not abandon it.

### From Jack

- Update setup docs to recommend MCP
- Rewrite skill to teach *how to use the MCP tools*, not curl
- Consider an **OpenClaw plugin** that bundles skill + simple tool wrapper
- Plugin could call Streamable HTTP / Remote MCP at `/mcp` endpoint
- Plugins support hooks — could auto-trigger recall at conversation start
- Maybe a guided MCP installer via skill file to smooth setup

### 1. MCP-first skill rewrite (Priority: High)

Rewrite `SKILL.md` to:
- Remove all curl commands and API reference
- Focus on **when/why** to use memory (behavioral rules, importance scoring, tagging)
- Reference the MCP tools (`mcp_memory_store_memory`, `mcp_memory_recall_memory`, etc.) instead of curl
- Keep the memory layers table and coexistence guidance for memory-core
- Add natural language examples: "If the user says 'remember X', store it. If they ask 'what do you know about X', recall it."

### 2. Update CLI to support MCP setup (Priority: High)

Extend `npx @verygoodplugins/mcp-automem openclaw` to:
- Add `--mode mcp|skill|plugin` flag (default: `mcp`)
- For MCP mode: configure mcporter entry or OpenClaw's MCP config to point at AutoMem
- For skill mode: keep current curl-based behavior as fallback
- Print appropriate post-install instructions per mode

### 3. OpenClaw plugin exploration (Priority: Medium)

Investigate building an OpenClaw plugin per https://docs.openclaw.ai/tools/plugin:
- Bundle the skill (behavioral rules) with a thin tool wrapper
- Tool wrapper calls AutoMem via Streamable HTTP at `/mcp` endpoint
- Use plugin hooks to auto-trigger recall at conversation start
- This would give us: native tools + behavioral guidance + auto-recall in one package

### 4. Natural language command mapping (Priority: Medium)

Add to skill behavioral rules:
- `remember <thing>` → `store_memory` with appropriate importance/tags
- `forget <thing>` → search + `delete_memory`
- `what do you know about <topic>` → `recall_memory`
- `update memory about <topic>` → recall + `update_memory`

### 5. Update setup docs (Priority: Low)

Rewrite `OPENCLAW_SETUP.md` to:
- Lead with MCP as recommended approach
- Document skill-only as alternative for users who prefer it
- Add plugin option once available
- Include troubleshooting for mcporter / MCP connection issues

## Non-Goals

- Removing skill support entirely (George-RD's feedback: some users prefer it)
- Auto-store without agent awareness (keep agent in the loop on what gets stored)
- Replacing memory-core — AutoMem should complement it, not replace it

## References

- OpenClaw plugin docs: https://docs.openclaw.ai/tools/plugin
- Current skill: `templates/openclaw/skill/SKILL.md`
- Current CLI: `src/cli/openclaw.ts`
- Setup guide: `templates/openclaw/OPENCLAW_SETUP.md`
- Streamable HTTP session fix: already landed (memory `e86e0dc3`)

### PR DESCRIPTION
**Pull request by jack-arturo**
## Summary

Brings OpenClaw to parity with the AutoMem integrations already shipping for Claude Desktop, Claude Code, Cursor, and Codex — specifically around using memory **beyond** the first-turn pre-seed. One-command install (`./install.sh`), non-technical-user friendly flow, and a shared memory-policy module that keeps all client templates in sync.

- **Plugin-first runtime**: native OpenClaw plugin registers typed AutoMem tools and injects a shared policy block via `prependSystemContext` (system prompt space — no leakage into the chat UI).
- **Shared policy surface** (`src/memory-policy/shared.ts`): first-turn preference recall (limit 20) + task recall (limit 30, 90-day window) + debug-only bugfix recall. Three mid-conversation store triggers (user correction / stabilized decision / articulated pattern) with explicit listener-cue phrases. Four-step atomic ritual: recall → store → verify → associate. Bare-tag discipline + 11 authorable relation types + never-store list.
- **One-command install** (`install.sh`): endpoint health-probe, memory-core explanation prompt, post-restart gateway health-poll with loud failure, final summary of what changed + doc link.
- **Cross-client parity tests**: snapshot + surface assertions validate that Claude Desktop, Claude Code session-start hook, Cursor template, Codex memory rules, and OpenClaw policy renderer all stay in lockstep.
- **Non-technical user docs**: new "What your AI is told" section in `OPENCLAW_SETUP.md` explains the policy block, the three storage triggers with example phrases, and the atomic ritual in plain language.

## Live-verified

Exercised on a fresh OpenClaw session against a populated AutoMem instance:

| Trigger | Expected | Result |
|---|---|---|
| First-turn explicit recall ("what do we know about X?") | Pre-seed + agent-initiated `automem_recall_memory` with clean args | ✅ |
| Stabilized decision ("ship it / final answer / lock it in") | Full 4-step ritual; `Decision`, importance 0.9 | ✅ |
| User correction ("actually, no — I prefer X / stop doing Y") | Full 4-step ritual; `Preference` w/ `correction` tag, `INVALIDATED_BY` against prior decision | ✅ |
| Routine follow-up (no trigger phrase) | No re-recall; uses conversation context | ✅ |
| Channel/group/hook sessions (`:group:`, `hook:`, exposure=off) | Policy stays in system prompt, not user bubble | ✅ (3 new regression tests cover this) |

## Tests

- `npx vitest run` → **223 passing** (14 suites)
- New coverage: memory-policy snapshot tests for all client templates including Codex, OpenClaw plugin regression tests for exposure/channel sessionKey edge cases.

## Files

- Core runtime: `src/openclaw-plugin.ts`, `src/memory-policy/shared.ts`, `src/openclaw-startup-profile.ts`, `src/automem-client.ts`
- CLI + install: `src/cli/openclaw.ts`, `install.sh`, `scripts/build-openclaw-plugin-package.mjs`
- Templates: `templates/{cursor,codex,openclaw}/`, `templates/claude-code/hooks/`, `templates/CLAUDE_DESKTOP_INSTRUCTIONS.md`
- Docs: `templates/openclaw/OPENCLAW_SETUP.md`, `INSTALLATION.md`

Fixes #68.

## Test plan

- [ ] Fresh clean install: `./install.sh` against a populated AutoMem endpoint → interactive prompts work, gateway restart completes healthy, summary prints.
- [ ] Fresh clean install against an *unreachable* endpoint → installer warns and re-prompts or exits cleanly.
- [ ] Send "let's ship X" on a fresh session → verify `store_memory(Decision) + associate_memories(PREFERS_OVER|REINFORCES)` fires.
- [ ] Send "actually no, I prefer Y" → verify `store_memory(Preference) + associate_memories(INVALIDATED_BY)` fires.
- [ ] Send a routine follow-up → verify no unnecessary recall.
- [ ] Confirm no `<automem-policy>` block appears in any chat bubble (DM, channel, or hook session).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/openclaw-plugin.ts [hunk 1]
```diff
@@ -102,14 +177,101 @@ function shouldAutoRecall(exposure: PluginConfig['exposure'], sessionKey?: strin
...
+    if (candidate) {
+      matches.add(candidate);
+    }
+  }
+
+  for (const match of source.matchAll(/\b[A-Z][a-z][A-Za-z0-9_-]{1,}\b/g)) {
+    const candidate = match[0]?.trim();
+    if (match.index === 0) {
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (jack-arturo) [reply src/openclaw-plugin.ts:250]: Keeping the position-0 skip to preserve the intent of the `does not auto-recall on ordinary follow-up turns` test (src/openclaw-plugin.test.ts:246) — removing the skip makes ordinary verbs like "Ship" become entities and fires a spurious recall. Users who need recall for a capitalized entity at the start of a prompt can trigger it explicitly via "what do you know about X?" (`EXPLICIT_RECALL_PROMPT_PATTERN` in src/memory-policy/shared.ts:28).

File: src/openclaw-plugin.ts [hunk 2]
```diff
@@ -9,14 +23,43 @@ import type {
...
   exposure: 'dm-only' | 'all' | 'off';
   defaultTags: string[];
+  startupProfile?: string;
 };
 
+type SessionState = {
+  seenEntities: Set<string>;
+};
+
+const sessionStates = new Map<string, SessionState>();
```
[the remark this answers is not in the document]
Comment (jack-arturo) [reply src/openclaw-plugin.ts:44]: Bounded both maps with simple LRU eviction — src/openclaw-plugin.ts:42 adds MAX_SESSION_STATES=500 and MAX_SEEN_ENTITIES_PER_SESSION=200, getSessionState re-inserts on hit to refresh LRU order, and a new addSeenEntity helper caps per-session entities. Skipped the full TTL class-hierarchy; the two caps address the unbounded-growth concern with much less code.

File: src/openclaw-plugin.ts [hunk 3]
```diff
@@ -370,52 +560,115 @@ const openClawPlugin = {
...
-      api.on('before_agent_start', async (event, ctx) => {
+      api.on('before_prompt_build', async (event, ctx) => {
         const prompt = String(event.prompt || '').trim();
+        const sections: string[] = [
+          renderOpenClawPolicyContext({ defaultTags: config.defaultTags }),
+        ];
+        const bootstrapSkipped = configSkipsBootstrap(api.config);
+        const startupTurn = bootstrapSkipped && isLikelyStartupTurn(event.messages);
+        const firstSubstantiveTurn =
+          isLikelyStartupTurn(event.messages) && isSubstantivePrompt(prompt);
```
[the remark this answers is not in the document]
Comment (jack-arturo) [reply src/openclaw-plugin.ts:601]: Parameterized the renderer with an optional limits override — src/memory-policy/shared.ts exports a new `OpenClawPolicyLimits` type, and the plugin now passes the active config values (`preferenceRecallLimit`, `contextRecallLimit`, `debugRecallLimit`, `contextRecallWindowDays`) into `renderOpenClawPolicyContext` at src/openclaw-plugin.ts:588, so the policy block always reflects runtime behavior.

File: src/memory-policy/shared.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+  'Pattern discovered -> Pattern -> EXEMPLIFIES',
+] as const;
+
+const AMBIGUOUS_PROJECT_TAGS = new Set(['api', 'app', 'test', 'video']);
+const CASUAL_OPENING_PATTERN =
+  /^(hi|hello|hey|yo|sup|thanks|thank you|ok|okay|cool|nice|great|ping|test|who are you)\b/i;
+const DEBUG_PROMPT_PATTERN =
+  /(error|exception|traceback|stack trace|stacktrace|failing|fails|failed|failure|bug|regression|crash|broken|debug|investigat|not work|doesn't work|does not work|cannot|can't|fix)/i;
+const EXPLICIT_RECALL_PROMPT_PATTERN =
+  /(what do (you|we) (have|know) about|tell me about|who is|who's|do you remember|remember|recall|search memory|check memory|look in memory|have we spoken about|what do you have on)/i;
```
[the remark this answers is not in the document]
Comment (jack-arturo) [reply src/memory-policy/shared.ts]: Tightened the regex at src/memory-policy/shared.ts:28 — dropped the bare `|remember|` alternative and added `what do you remember about` / `remember when` so store-intent phrasing ("remember that I prefer X") no longer forces a recall phase. Kept `do you remember` since that's unambiguously a recall request.

File: INSTALLATION.md [hunk 5]
```diff
@@ -898,10 +900,24 @@ AutoMem now supports three OpenClaw integration modes, in this recommended order
...
+   Local-build equivalent while developing this repo:
+
+   ```bash
+   ./install.sh
+   ```
+
+4. **Open OpenClaw**. The installer will restart the gateway, verify the plugin, and open or print the dashboard URL.
+
+`openclaw --mode plugin` installs a lean native plugin package staged inside `@verygoodplugins/mcp-automem`. That keeps `memory-core` as OpenClaw's active memory slot instead of replacing it.
+It also enables the `/plugins` chat command so users can verify and inspect the install from inside OpenClaw.
```
[the remark this answers is not in the document]
Comment (jack-arturo) [reply INSTALLATION.md]: Clarified at INSTALLATION.md:914 — replaced the bare `openclaw --mode plugin` with "The recommended setup command above installs OpenClaw in `plugin` mode using the packaged flow from `@verygoodplugins/mcp-automem` (under the hood: `npx @verygoodplugins/mcp-automem openclaw --mode plugin`)" so there's no ambiguity that `openclaw` itself doesn't accept that flag.

File: install.sh [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+    ;;
+  *)
+    echo "Unsupported AutoMem client: $selected_client" >&2
+    exit 1
+    ;;
+esac
+
+echo "Restarting OpenClaw gateway..."
+if ! node -e 'const { spawnSync } = require("child_process"); const result = spawnSync("openclaw", ["gateway", "restart"], { stdio: "ignore", timeout: 20000 }); if (result.error && result.error.code !== "ETIMEDOUT") process.exit(1); process.exit(result.status ?? 1);' >/dev/null 2>&1; then
+  echo "OpenClaw gateway restart did not complete cleanly. Retry manually: openclaw gateway restart" >&2
```
[the remark this answers is not in the document]
Comment (jack-arturo) [reply install.sh]: Made ETIMEDOUT a soft failure — install.sh:249 now prints a warning and exits 0 on timeout so `wait_for_gateway_health` (the health-poll loop that follows) decides pass/fail based on actual gateway health. Real spawn errors and non-zero exit codes still abort loudly.
